### PR TITLE
Generated mock txs do not result in a net negative balance

### PR DESCRIPTION
**Pull request by sarthaktexas**
fixes #4438. This also adds mock fees to help the onboarding team show how fees are represented on the platform.

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -143,53 +144,104 @@ def show
...
+            receipts: Array.new(rand(100) < 90 ? 1 : 0), # 90% chance of 1 receipt, 10% chance of no receipts
+            comments: Array.new(rand(9) > 1 ? 0 : rand(1..2)), # 1/3 chance of no comments, 2/3 chance of 1 or 2 comments
             donation?: !trans[:amount].negative?,
             donation: !trans[:amount].negative? ? nil : OpenStruct.new(recurring?: trans[:monthly]),
             tags: []
           )
         )
       end
+      puts @transactions
+
```
Comment (sampoder) [inline app/controllers/events_controller.rb]:
```suggestion

```

probs best to remove this, PR looks sick!

Comment (sarthaktexas) [reply app/controllers/events_controller.rb]: ah yes marked as draft due to these logs

**Comment by sarthaktexas** [comment]
~~last thing this needs (apart from Robocop linting) is making the fiscal sponsorship fees actually a fiscal sponsorship fee type so it displays properly~~ nvm thanks sam :)

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -143,53 +144,101 @@ def show
...
+      @transactions.clear # clear the transactions array (only really matters for development testing)
+
+
+      @mock_tx.each do |trans|
         @transactions << OpenStruct.new(
           amount: trans[:amount],
-          amount_cents: rand(1000),
+          amount_cents: trans[:amount] * 100,
           fee_payment: true,
```
Comment (sampoder) [inline app/controllers/events_controller.rb]:
```suggestion
          fee_payment: true,
          fee_payment?: trans[:desc].include?("💰 Fiscal sponsorship fee") ? true : false,
```

@sarthaktexas this should do it!

**Comment by garyhtou** [comment]
I haven't taken a close look at the code yet, however, I just wanted to throw out there that it'd be nice to move this mock data generation into a service! (Myself, Sam, or anyone else would be happy to help show you how to do that)

**Comment by sampoder** [comment]
I’m down to help!

**Comment by sarthaktexas** [comment]
now it's ready for review :)

**Review by sampoder — APPROVED** [review]
looks good to me!

File: app/services/mock_transaction_engine_service/generate_mock_transaction.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        )
+      end
+    end
+
+    def generate_mock_tx
+      if @mock_balance < 0
+        raise "Balance is less than 0. Something's gone wrong."
+      else
+        NEGATIVE_DESCRIPTIONS[rand(NEGATIVE_DESCRIPTIONS.length)].merge({ amount: rand(0..@mock_balance) * -1 })
+      end
```
Comment (garyhtou) [inline app/services/mock_transaction_engine_service/generate_mock_transaction.rb]: IIRC, you had mentioned that this should never happen. let's remove it then!

```suggestion
      NEGATIVE_DESCRIPTIONS[rand(NEGATIVE_DESCRIPTIONS.length)].merge({ amount: rand(0..@mock_balance) * -1 })
```

**Review by garyhtou — APPROVED** [review]
yayyy! excited to see that we've got non-negative balances now :)

**Pull request merged**